### PR TITLE
Added TransposeReshapeEliminationForMatmul Transformation

### PR DESCRIPTION
**Pull request by a-sidorova**
### Details:
 - *Added Transformation which eliminates movement operations (Reshape and Transpose) near MatMul because MatMul supports inputs with different shapes. There is support only for cases when 2nd input of MatMul is contracted!*
 - *Added in CommonOptimizations to other plugins can use it*
 - *Not supported dynamic shapes*
 - *DLB results are attached to the ticket*

Examples:

For TF1 model with no einsum support
image

For TF2 model with einsum (IR with einsum)
image

### Tickets:
 - *68710*

### CODE REVIEW

File: inference-engine/src/transformations/src/transformations/common_optimizations/transpose_reshape_elimination_for_matmul.cpp [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+
+#include <memory>
+#include <vector>
+
+#include <ngraph/opsets/opset1.hpp>
+#include <ngraph/rt_info.hpp>
+#include <ngraph/pattern/op/wrap_type.hpp>
+#include <ngraph/validation_util.hpp>
+#include "itt.hpp"
+
```
[the remark this answers is not in the document]
Comment (a-sidorova) [reply inference-engine/src/transformations/src/transformations/common_optimizations/transpose_reshape_elimination_for_matmul.cpp:15]: Added, thanks!

File: inference-engine/src/mkldnn_plugin/mkldnn_plugin.cpp [hunk 2]
```diff
@@ -333,6 +334,7 @@ static void TransformationUpToCPUSpecificOpSet(std::shared_ptr<ngraph::Function>
     pass_config->enable<ngraph::pass::ConvertInterpolate1ToInterpolate4>();
     pass_config->enable<ngraph::pass::ConvertGather1ToGather7>();
     pass_config->enable<ngraph::pass::ConvertGather8ToGather7>();
+    pass_config->enable<ngraph::pass::TransposeReshapeEliminationForMatmul>();
```
Comment (rkazants) [inline inference-engine/src/mkldnn_plugin/mkldnn_plugin.cpp:337]: why do we activate the transformation in the plugin as well? I think the common optimization pass is sufficient.

**Review by rkazants — APPROVED** [review]
please call it only in common optimization pass

**Comment by rkazants** [comment]
@ilyachur, @GlebKazantaev, please review it.

File: inference-engine/src/transformations/src/transformations/common_optimizations/transpose_reshape_elimination_for_matmul.cpp [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+        auto transpose_after_constant = std::dynamic_pointer_cast<ngraph::opset1::Constant>(transpose_after->get_input_node_shared_ptr(1));
+        if (!transpose_before || !transpose_after || !transpose_before_constant || !transpose_after_constant)
+            return false;
+        auto transpose_before_order = transpose_before_constant->cast_vector<int64_t>();
+        auto transpose_after_order = transpose_after_constant->cast_vector<int64_t>();
+        // need to check that input shape is correctly contracted and output shape is correctly unpacked using transposes
+        if (!check_transposes(transpose_before_order, transpose_after_order, transposed_b))
+            return false;
+
+        const auto new_matmul = std::make_shared<opset1::MatMul>(input_1, input_2, transposed_a, false);
```
Comment (rkazants) [inline inference-engine/src/transformations/src/transformations/common_optimizations/transpose_reshape_elimination_for_matmul.cpp:170]: is transpose_b for new MatMul always False?

Comment (rkazants) [reply inference-engine/src/transformations/src/transformations/common_optimizations/transpose_reshape_elimination_for_matmul.cpp:170]: discussed offline. it is correct

File: inference-engine/src/mkldnn_plugin/mkldnn_plugin.cpp (hunk 2 above)
Comment (a-sidorova) [reply inference-engine/src/mkldnn_plugin/mkldnn_plugin.cpp]: Enabled for all plugins as default. Thanks!

File: inference-engine/src/transformations/src/transformations/common_optimizations/common_optimizations.cpp [hunk 4]
```diff
@@ -150,6 +151,9 @@ bool ngraph::pass::CommonOptimizations::run_on_function(std::shared_ptr<ngraph::
     // CF is required after all decompositions
     manager.register_pass<ngraph::pass::ConstantFolding>();
 
+    // TransposeReshapeEliminationForMatmul must be after EinsumDecomposition
+    manager.register_pass<ngraph::pass::TransposeReshapeEliminationForMatmul>();
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/src/transformations/common_optimizations/common_optimizations.cpp:155]: You can put this pass together with other passes in decomp. So no additional graph traversals will be performed.

File: inference-engine/src/transformations/src/transformations/common_optimizations/transpose_reshape_elimination_for_matmul.cpp [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+    auto const_transpose_after_pattern = ngraph::pattern::wrap_type<opset1::Constant>();
+    auto transpose_after_pattern = ngraph::pattern::wrap_type<opset1::Transpose>({reshape_after_pattern, const_transpose_after_pattern});
+
+    ngraph::matcher_pass_callback callback = [=](pattern::Matcher& m) {
+        const auto& pattern_value_map = m.get_pattern_value_map();
+        const auto& input_1 = pattern_value_map.at(input_1_pattern);
+        const auto& input_2 = pattern_value_map.at(input_2_pattern);
+
+        // this transformation supports only static shapes
+        if (input_1.get_partial_shape().is_dynamic() || input_2.get_partial_shape().is_dynamic())
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/src/transformations/common_optimizations/transpose_reshape_elimination_for_matmul.cpp:133]: Please move this checks to pattern predicates.

File: inference-engine/src/transformations/src/transformations/common_optimizations/transpose_reshape_elimination_for_matmul.cpp [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+                pattern_value_map.at(const_reshape_before_pattern).get_node_shared_ptr());
+        if (!reshape_before || !reshape_after || !reshape_before_constant)
+            return false;
+        if (!check_input_reshape(reshape_before, reshape_before_constant->cast_vector<int64_t>(), transposed_b))
+            return false;
+
+        // this transformation supports only cases when 2nd input of matmul is transpose->reshape
+        if (matmul->get_input_node_shared_ptr(1) != reshape_before)
+            return false;
+        if (matmul->get_input_shape(0).size() != 2)
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/src/transformations/common_optimizations/transpose_reshape_elimination_for_matmul.cpp:154]: Please move to pattern predicate.

File: inference-engine/tests/functional/inference_engine/transformations/transpose_reshape_elimination_for_matmul.cpp [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+#include <ngraph/pass/manager.hpp>
+
+#include "common_test_utils/ngraph_test_utils.hpp"
+
+
+using namespace testing;
+using namespace ngraph;
+
+
+TEST(TransformationTests, TransposeReshapeEliminationForMatMul) {
```
Comment (GlebKazantaev) [inline inference-engine/tests/functional/inference_engine/transformations/transpose_reshape_elimination_for_matmul.cpp:24]: Please use TEST_F and TransformationsTestF fixture.

**Comment by a-sidorova** [comment]
@GlebKazantaev @jane-intel I applied comments. Take a look please

**Comment by andrei-kochin** [comment]
@jane-intel please take a look

**Review by dmitry-gorokhov — APPROVED** [review]

**Comment by a-sidorova** [comment]
@dmitry-gorokhov DLB passed, CI is green. I think, We can merge :blush:

**Comment by andrei-kochin** [comment]
@GlebKazantaev please take a look

File: src/common/transformations/src/transformations/common_optimizations/common_optimizations.cpp [hunk 8]
```diff
@@ -51,6 +51,7 @@
 #include "transformations/common_optimizations/mul_conv_fusion.hpp"
 #include "transformations/common_optimizations/interpolate_sequence_fusion.hpp"
 #include "transformations/common_optimizations/convert_compression_only_to_legacy.hpp"
+#include <transformations/common_optimizations/transpose_reshape_elimination_for_matmul.hpp>
```
Comment (GlebKazantaev) [inline src/common/transformations/src/transformations/common_optimizations/common_optimizations.cpp:54]: nit: 
```suggestion
#include "transformations/common_optimizations/transpose_reshape_elimination_for_matmul.hpp"
```

**Review by GlebKazantaev — APPROVED** [review]

**Pull request merged**